File: test-suite/src/test/java/io/micronaut/crac/CheckpointSimulatorTest.java
package io.micronaut.crac;

import io.micronaut.context.ApplicationContext;
import io.micronaut.context.env.Environment;
import io.micronaut.crac.test.CheckpointSimulator;
import io.micronaut.runtime.server.EmbeddedServer;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@MicronautTest
public class CheckpointSimulatorTest {

    //tag::test[]
    @Test
    void emulateCheckpoint() {
        try (EmbeddedServer server = ApplicationContext.run(EmbeddedServer.class, Environment.TEST)) {
            CheckpointSimulator checkpointSimulator = server.getApplicationContext().getBean(CheckpointSimulator.class);  // <1>
            assertTrue(server.isRunning());

            checkpointSimulator.runBeforeCheckpoint(); // <2>
            assertFalse(server.isRunning());

            checkpointSimulator.runAfterRestore(); // <3>
            assertTrue(server.isRunning());
        }
    }
    //end::test[]
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/crac/CheckpointSimulatorTest.kt
package io.micronaut.crac

import io.micronaut.context.ApplicationContext
import io.micronaut.context.env.Environment
import io.micronaut.crac.test.CheckpointSimulator
import io.micronaut.runtime.server.EmbeddedServer
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test


@MicronautTest
class CheckpointSimulatorTest {

    //tag::test[]
    @Test
    fun emulateCheckpoint() {
        ApplicationContext.run(EmbeddedServer::class.java, Environment.TEST).use { server ->
            val checkpointSimulator = server.applicationContext.getBean( // <1>
                CheckpointSimulator::class.java
            )
            Assertions.assertTrue(server.isRunning)

            checkpointSimulator.runBeforeCheckpoint() // <2>
            Assertions.assertFalse(server.isRunning)

            checkpointSimulator.runAfterRestore() // <3>
            Assertions.assertTrue(server.isRunning)
        }
    }
    //end::test[]
}

